### ISSUE TITLE
Unused CSS gives useless URL

### ISSUE THREAD

**Issue by Pascal666** [GoogleChrome/lighthouse#5626]
#### Provide the steps to reproduce
1. Run LH on http://giveawaylisting.com/ or http://giveawaylisting.com/index2.html

#### What is the current behavior?
Unused CSS gives a useless URL
<img width="515" alt="lighthouse" src="https://user-images.githubusercontent.com/1330708/42409166-ea203bb2-819b-11e8-9ae0-6d8cd7c528d6.png">

#### What is the expected behavior?
Give me some way to actually find the unused CSS so maybe I can remove it.

#### Environment Information
* Affected Channels: DevTools
* Lighthouse version: 2.9.1
* Node.js version:
* Operating System: Windows 7

**Related issues**

**Comment by patrickhulce** [comment]
That's because there is no URL for these, they were inline stylesheets. The "URL" we're giving you is the content of the inline style because that's the only way we have to identify it.

**Comment by Pascal666** [comment]
Inline or not, it had to have come from some file.  Without the URL of that file, the warning is nearly useless.  That CSS is not inline in the original URL.  If there is some other way to find the URL where that CSS came from, it needs to be documented in the Learn more link.

**Comment by patrickhulce** [comment]
> If there is some other way to find the URL where that CSS came from

We don't, but the developer does. Inline stylesheets could've been generated dynamically via script, shipped as a string, etc. The content is all we have to identify it and hopefully that's enough for the developer to know where to start looking. For example, I know absolutely nothing about that site, but I can already guess that it's an angular stylesheet from a [single search](https://github.com/search?q=.ng-hide%3Anot%28.ng-hide-animate%29+%7B+display%3A+none&type=Code).

We recognize it's not ideal, but it's not a very debuggable way of shipping CSS to begin with. I'm not sure what you'd recommend we do instead when there's no attributable URL. The best I see is more clearly marking it as an inline style for the dev to investigate.

**Comment by Pascal666** [comment]
The developer of giveawaylisting.com?  That's me.  I have no idea where that CSS is coming from.  That's why I opened this Issue.

You saying it was probably angular allowed me to find it though.  I hit F12, selected the network tab, and looked for any filenames containing angular.  The below is the URL that contains the CSS in question.  I would have expected LH to give me this URL.

https://ajax.googleapis.com/ajax/libs/angularjs/1.3.5/angular.min.js

**Comment by patrickhulce** [comment]
The process to find these is not exact and requires general intuition which is why LH just shows you the content, when you inspect the page you just see an inline style you don't know how it got there. If you would like to document how to debug pages to find out where inline stylesheets are coming from, I'm sure the community would be grateful :)

FWIW, it's not actually from that file, the stylesheet added by angular.min.js is not 109 KB. The one LH is flagging is `<style id="legacy-shr-base-styles">`, which seems to be added by`http://apps.shareaholic.com/v2/963db2d7366f054efaaad62cfbbfd73142390d18/shrMain.min.js` which likely uses angular under the hood.

Steps I took to figure this out:

1. Ran `[...document.querySelectorAll('style')].filter(el => el.textContent.includes('.ng-hide:not(.ng-hide-animate)'))[0].textContent.slice(0, 1000)` in the console
2. Notice that the stylesheet references "shareaholic" quite a bit
3. Look at shareaholic requests in the network panel

**Comment by Pascal666** [comment]
Thank you.  Are you saying LH simply queries the DOM, it does not look at the actual files as they are loaded?

Considering that I found the wrong file, and don't yet understand what you did to figure it out, I'm probably not qualified to write those instructions.  I was actually kinda hoping it was Shareaholic.  They are on my "replace soon" list anyway.  Don't suppose you can recommend another package for adding social media icons to a web page?

**Comment by paulirish** [comment]
it'd be some work but we could indicate the root document's URL and say  `(inline)` in the report proceeding this selector.

**Comment by connorjclark** [comment]
My work in `font-size` #9356 could be done to improve attribution here too.

**Comment by connorjclark** [comment]
For inline CSS within an HTML document, we could do one of these things:

1. Use a `node` detail type - this may provide enough data for a developer to determine where to look. In DevTools, it would link to the node in the Element panel.
1. Use a `source-location` detail type - this will give an exact line/col in the HTML where the CSS rule starts. In DevTools, it would link directly to that position of the HTML in the Sources panel.
1. Use a `snippet` detail type - this is the only one that could keep the "content preview". There is no linking in DevTools for this detail item - if we decide to use snippets more, I'd like to add support for DevTools to link to the Sources panel to the code its showing.

**Comment by patrickhulce** [comment]
I'm concerned about `node` and `source-location` losing value and portability when not viewed in devtools. The snippet is the only useful type IMO when viewed in the standalone report, PSI, etc.

Would `node` provide any identification other than just `<style>` for the webpack-style CSS?

I have a feeling that `source-location` wouldn't be able to actually pinpoint the JS file that contains the content, but *if* we could that would be pretty awesome.

**Comment by connorjclark** [comment]
> I'm concerned about node and source-location losing value and portability when not viewed in devtools. The snippet is the only useful type IMO when viewed in the standalone report, PSI, etc.

yup, so I would like to tack on the info needed for linkifying to the `snippet` item type. looks like it'd just be adding `url` to the item (and I guess `urlProvider` ...)

> Would node provide any identification other than just <style> for the webpack-style CSS?

just the `style`.

> I have a feeling that source-location wouldn't be able to actually pinpoint the JS file that contains the content, but if we could that would be pretty awesome.

JS won't be considered here. `source-location` would point to the exact rule declaration within an inline (or external) stylesheet.

If we want to know a style is created by JS we could use [DOM.getNodeStackTraces](https://chromedevtools.github.io/devtools-protocol/tot/DOM#method-getNodeStackTraces).

**Comment by connorjclark** [comment]
https://github.com/GoogleChrome/lighthouse/blob/retry-some-jest-tests/lighthouse-core/computed/unused-css.js#L129

We can create the `source-location` data from `CSSStyleSheetInfo['header']`.

Instead of augmenting `snippet` to linkify, what if we gave `source-location` a fallback text to show outside of DevTools? Or, a text to always show (in combination with the URL:line:col)?

**Comment by connorjclark** [comment]
There's still value in doing this. We need a fallback text (see above comment) for source-location (DevTools RPP does this sort of thing for the same reasons)

**Comment by Pascal666** [comment]
It has been nearly 7 years since this ticket was opened. The website mentioned in the original bug report no longer exists as it did at the time. The below links to a copy of the page preserved at the Internet Archive:

https://web.archive.org/web/20180703032624/http://www.giveawaylisting.com/
